<div class="alert alert-block alert-success" style="font-family: Times New Roman">
    <h4><strong>Laboratory Task 2</strong></h4>

<p style="font-family:Times New Roman; text-align:justify; font-size:15px">
    <b>Instruction:</b> Perform a single forward pass and compute for the error.
</p>

$$x = \begin{bmatrix} 1 \\ 0 \\ 1 \end{bmatrix}$$

$$y = \begin{bmatrix} 1 \end{bmatrix}$$

$$f=max(0, Z_n)$$

$$\text{hidden unit weights} = 
\begin{bmatrix}
w_{11} = 0.2 && w_{12} = -0.3 \\
w_{13} = 0.4 && w_{14} = 0.1 \\
w_{15} = -0.5 && w_{16} = 0.2
\end{bmatrix}$$

$$\text{output unit weights} = 
\begin{bmatrix}
w_{21} = -0.3 \\
w_{22} = -0.2
\end{bmatrix}$$

$$\theta = 
\begin{bmatrix}
\theta_{1} = -0.4 \\
\theta_{2} = 0.2 \\
\theta_{3} = 0.1
\end{bmatrix}$$
</div>

## Network Setup

The network has **3 input units**, **2 hidden units**, and **1 output unit**.

- $w_{11}, w_{13}, w_{15}$ connect inputs $x_1, x_2, x_3$ to **hidden unit 1**
- $w_{12}, w_{14}, w_{16}$ connect inputs $x_1, x_2, x_3$ to **hidden unit 2**
- $w_{21}, w_{22}$ connect hidden units 1 and 2 to the **output unit**
- $\theta_1, \theta_2$ are the biases for the hidden units, $\theta_3$ is the bias for the output unit
- Activation function: $f(Z) = \max(0, Z)$ (ReLU)

For any unit $j$, the net input is:

$$Z_j = \sum_i w_{ij} \, O_i + \theta_j$$

and the unit's output is $O_j = f(Z_j) = \max(0, Z_j)$.

In [1]:
import numpy as np

# Input vector
x = np.array([1, 0, 1])

# Target output
y = 1

# Hidden layer weights (rows = inputs x1,x2,x3 ; columns = hidden unit 1, hidden unit 2)
W_hidden = np.array([
    [0.2, -0.3],   # w11, w12  (from x1)
    [0.4,  0.1],   # w13, w14  (from x2)
    [-0.5, 0.2]    # w15, w16  (from x3)
])

# Output layer weights (hidden unit 1 -> output, hidden unit 2 -> output)
W_output = np.array([-0.3, -0.2])

# Biases: theta1 (hidden 1), theta2 (hidden 2), theta3 (output)
theta1, theta2, theta3 = -0.4, 0.2, 0.1

def relu(z):
    return np.maximum(0, z)

print("x =", x)
print("y =", y)
print("W_hidden =\\n", W_hidden)
print("W_output =", W_output)
print("theta1, theta2, theta3 =", theta1, theta2, theta3)

x = [1 0 1]
y = 1
W_hidden =\n [[ 0.2 -0.3]
 [ 0.4  0.1]
 [-0.5  0.2]]
W_output = [-0.3 -0.2]
theta1, theta2, theta3 = -0.4 0.2 0.1


## Step 1: Net Input to the Hidden Layer

$$Z_1 = x_1 w_{11} + x_2 w_{13} + x_3 w_{15} + \theta_1$$
$$Z_2 = x_1 w_{12} + x_2 w_{14} + x_3 w_{16} + \theta_2$$

In [2]:
theta_hidden = np.array([theta1, theta2])

Z_hidden = x @ W_hidden + theta_hidden

print("Z_hidden (net input to hidden units) =", Z_hidden)

Z_hidden (net input to hidden units) = [-0.7  0.1]


## Step 2: Hidden Layer Activation (ReLU)

$$O_1 = f(Z_1) = \max(0, Z_1) \qquad O_2 = f(Z_2) = \max(0, Z_2)$$

In [3]:
O_hidden = relu(Z_hidden)

print("O_hidden (hidden unit outputs) =", O_hidden)

O_hidden (hidden unit outputs) = [0.  0.1]


## Step 3: Net Input to the Output Layer

$$Z_3 = O_1 w_{21} + O_2 w_{22} + \theta_3$$

In [4]:
Z_output = O_hidden @ W_output + theta3

print("Z_output (net input to output unit) =", Z_output)

Z_output (net input to output unit) = 0.08


## Step 4: Output Layer Activation (ReLU)

$$\hat{y} = f(Z_3) = \max(0, Z_3)$$

In [5]:
y_hat = relu(Z_output)

print("Predicted output (y_hat) =", y_hat)

Predicted output (y_hat) = 0.08


## Step 5: Compute the Error

Using the simple error (target minus predicted output):

$$\text{Error} = y - \hat{y}$$

and the squared error (a common loss for a single sample):

$$\text{Squared Error} = \frac{1}{2}(y - \hat{y})^2$$

In [6]:
error = y - y_hat
squared_error = 0.5 * (y - y_hat) ** 2

print("Error (y - y_hat)     =", error)
print("Squared Error         =", squared_error)

Error (y - y_hat)     = 0.92
Squared Error         = 0.4232


## Summary of Results

| Quantity | Value |
|---|---|
| $Z_1$ (hidden unit 1 net input) | -0.7 |
| $Z_2$ (hidden unit 2 net input) | 0.1 |
| $O_1$ (hidden unit 1 output) | 0.0 |
| $O_2$ (hidden unit 2 output) | 0.1 |
| $Z_3$ (output unit net input) | 0.08 |
| $\hat{y}$ (predicted output) | 0.08 |
| Error ($y - \hat{y}$) | 0.92 |
| Squared Error | 0.4232 |

Since hidden unit 1's net input ($Z_1 = -0.7$) is negative, ReLU zeroes it out completely —
it contributes nothing to the output. Only hidden unit 2 passes a (small) positive signal
forward, so the final prediction $\hat{y} = 0.08$ is far from the target $y = 1$, giving a
fairly large error.